scripts/generate-lint.go

```diff
@@ -0,0 +1,167 @@
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path"
+	"strings"
+
+	"github.com/pkg/errors"
+)
+
+const (
+	lintPrefix        = "lint"
+	lintVariant       = "race-detector"
+	lintGroup         = "lint-group"
+	groupMaxHosts     = 1
+	evergreenLintTask = "evergreen"
+	jsonFilename      = "generate-lint.json"
+)
+
+// whatChanged returns a list of files that have changed in the working
+// directory. First, it tries diffing changed files against the merge base. If
+// there are no changes, this is not a patch build, so it diffs HEAD against HEAD~.
+func whatChanged() ([]string, error) {
+	mergeBaseCmd := exec.Command("git", "merge-base", "master@{upstream}", "HEAD")
+	base, err := mergeBaseCmd.Output()
+	if err != nil {
+		return nil, errors.Wrap(err, "problem getting merge-base")
+	}
+	diffCmd := exec.Command("git", "diff", strings.TrimSpace(string(base)), "--name-only")
+	files, err := diffCmd.Output()
+	if err != nil {
+		return nil, errors.Wrap(err, "problem getting diff")
+	}
+	var split []string
+	// if there is no diff, this is not a patch build, so diff against HEAD~
+	if len(files) == 0 {
+		diffCmd := exec.Command("git", "diff", "HEAD~", "--name-only")
+		files, err = diffCmd.Output()
+		if err != nil {
+			return nil, errors.Wrap(err, "problem getting diff")
+		}
+	}
+	split = strings.Split(strings.TrimSpace(string(files)), "\n")
+	return split, nil
+}
+
+// targetsFromChangedFiles returns a list of make targets.
+func targetsFromChangedFiles(files []string) ([]string, error) {
+	targets := map[string]struct{}{}
+	for _, f := range files {
+		filePath := strings.TrimSpace(f)
+		if strings.HasSuffix(filePath, ".go") {
+			dir := path.Dir(filePath)
+			if dir == "scripts" {
+				continue
+			}
+			if dir == "." || dir == "main" {
+				targets["evergreen"] = struct{}{}
+			} else {
+				targets[strings.Replace(dir, "/", "-", -1)] = struct{}{}
+			}
+		}
+	}
+	targetSlice := []string{}
+	for t := range targets {
+		targetSlice = append(targetSlice, t)
+	}
+	return targetSlice, nil
+}
+
+// makeTask returns a task map that can be marshaled into a JSON document.
+func makeTask(target string) map[string]interface{} {
```

Comment:
it might be reasonable to package these task generation functions up into some sort of externalizeable library for actual users' use. 

worth a ticket, early next quarter?